src/lseq.rs

```diff
@@ -133,71 +141,59 @@ impl<T: Clone, A: Ord + Clone> LSeq<T, A> {
     /// # Panics
     ///
     /// * If the allocation of a new index was not between `ix` and `ix - 1`.
-    pub fn insert_index(&mut self, ix: usize, val: T) -> Op<T, A> {
-        let min_id = self.gen.lower();
-        let max_id = self.gen.upper();
-
-        // If we're inserting past the length of the LSEQ then it's the same as appending.
-        let (lower_id, upper_id) = if self.seq.len() <= ix {
-            let prev = self.seq.last().map(|entry| &entry.id).unwrap_or(&min_id);
-            (prev, &max_id)
-        } else {
-            // To insert an element at position ix, we want it to appear in the sequence between
-            // ix - 1 and ix. To do this, retrieve each bound defaulting to the lower and upper
-            // bounds respectively if they are not found.
-            let prev = match ix.checked_sub(1) {
-                Some(i) => &self.seq[i].id,
-                None => &min_id,
-            };
-            let next = self.seq.get(ix).map(|entry| &entry.id).unwrap_or(&max_id);
-
-            (prev, next)
+    pub fn insert_index(&mut self, mut ix: usize, val: T, actor: A) -> Op<T, A> {
+        ix = ix.min(self.seq.len());
+        let zero = BigRational::zero();
+        let one = BigRational::one();
+        let two = BigRational::from_integer(2.into());
+
+        // TODO: replace this logic with BTreeMap::range()
+        let (prev, next) = match ix.checked_sub(1) {
+            Some(indices_to_drop) => {
+                let mut indices = self.seq.keys().skip(indices_to_drop);
+                (indices.next(), indices.next())
+            }
+            None => {
+                // Inserting at the front of the list
+                let mut indices = self.seq.keys();
+                (None, indices.next())
+            }
         };
 
-        let ix_ident = self.gen.alloc(&lower_id, &upper_id);
-
-        assert!(lower_id < &ix_ident);
-        assert!(&ix_ident < upper_id);
-
-        Op::Insert {
-            id: ix_ident,
-            dot: self.clock.inc(self.actor()),
-            val,
-        }
+        let id = match (prev, next) {
+            (Some(p), Some(q)) => (&p.id + &q.id) / two,
+            (Some(p), None) => &p.id + &one,
+            (None, Some(p)) => &p.id - &one,
```

Comment:
Perhaps for more readability make it `(None, Some(q))`?

Reply:
Yes, good point